### ISSUE TITLE
hsa-rocr-dev dies looking for version of hsakmt-roct in all spack build

### ISSUE THREAD

**Issue by mwkrentel** [LLNL/spack#29381]
### Steps to reproduce the issue

ping @srekolam @arjun-raj-kuppala @haampie (maintainers)

For me, the all-spack build of `hip` or `hsa-rocr-dev` fails as:

```
% spack install hsa-rocr-dev
...
     16    -- Performing Test ELF_GETSHDRSTRNDX
     17    -- Performing Test ELF_GETSHDRSTRNDX - Success
  >> 18    CMake Error at CMakeLists.txt:97 (find_package):
     19      Could not find a configuration file for package "hsakmt" that is compatible
     20      with requested version "1.0".
     21    
     22      The following configuration files were considered but not accepted:
     23    
     24        /home/krentel/elf/install/linux-rhel8-zen/gcc-8.5.0/hsakmt-roct-4.5.2-2okqlbgb6jwwvb5toxejfgayygl3jjth/lib64/cmake/hsakmt/hsakmt-config.cmake, version: 0.11.2-16733-gd424d0e54e
```

This happens with 4.5.2 and 4.3.1.  I'm using an all-spack build, that
is no `packages.yaml` for any external packages, and not using `/opt/rocm`.

Indeed, `CMakelists.txt` in hsa-rocr-dev has:

```
  95 ## Find external dependencies.
  96 find_package(LibElf REQUIRED)
  97 find_package(hsakmt 1.0 REQUIRED HINTS ${CMAKE_INSTALL_PREFIX} PATHS /opt/rocm)
```

And `/opt/rocm` has this in `hsakmt-config-version.cmake` which is certainly > 1.0:

```
set(PACKAGE_VERSION "20210902.12.3277-mainline-20210902-12-g3277d53")
```

But if you build `hsakmt-roct` from scratch in spack, you get this which is < 1.0:

```
set(PACKAGE_VERSION "0.11.2-16011-g186730a284")
```

So, it looks to me that you can't build hip or hsa-rocr-dev from
scratch in spack.

Thoughts on the best way to fix this?
Patch the CMakeLists.txt test?
Patch the hsakmt-config-version.cmake file?

This is more than the most recent 4.5.2, it's also a problem in 4.3.1
and probably several (all?) revs earlier.

### Error message

<details><summary>Error message</summary><pre>
...
</pre></details>

### Information on your system

$ spack debug report
* **Spack:** 0.11.2-16011-186730a284
* **Python:** 3.6.8
* **Platform:** linux-rhel8-zen2
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

**Comment by srekolam** [comment]
@mwkrentel  can you let me know the steps to reproduce ..i am not seeing this issue - i looked at the hsakmt-config-version.cmake inside the install_prefix_path of hsakmt-roct/lib64/cmake/hsakmt i see the below set(PACKAGE_VERSION "1.0.0")
I looked at hsakmt-roct@5.0.0 as well as hsakmt-roct@4.5.2 too

**Comment by mwkrentel** [comment]
@srekolam Hmm, we must be doing something different, but at first
glance, I don't see what it is.

For me, this is an ALL-spack build.  That is, NO external packages
from `packages.yaml`.  Nothing `spack external find`.  Also, you can't
use rocm 5.0 because that's not in spack yet.

I'm on x86_64, RH 8.5, gcc 8.5.0.  I could try other arch's, but I
don't expect that to make any difference.

I'm mostly using 4.5.2, but it also happens to me with 4.3.1.

There is /opt/rocm on this machine, but I don't think anything is
using it.  Certainly, it shouldn't.

Do you have any idea where version 0.11.2-16733-gd424d0e54e-dirty
would come from?

set(PACKAGE_VERSION "0.11.2-16733-gd424d0e54e-dirty")

I don't know.  Double check your packages.yaml and spack.yaml files.

**Comment by haampie** [comment]
@mwkrentel with `bwrap --dev-bind / / --tmpfs /opt/rocm spack install ...` you can make sure `/opt/rocm` is not used.

**Comment by mwkrentel** [comment]
@srekolam  @haampie 
I tried this on another machine that doesn't have GPUs, so no
/opt/rocm files and no chance of using them.

I get the same result.  Again, with NO external packages, hsakmt-roct produces:

```
hsakmt-config-version.cmake:

set(PACKAGE_VERSION "0.11.2-16735-g5bc518baf9")
```

Is this an issue with cmake (3.22.2) or pkgconf (1.8.0) ?

@srekolam   Have you tried building hsakmt-roct with NO external packages,
no `packages.yaml`, no `spack.yaml`, no environment ?

For me, it's enough to run this and then inspect `hsakmt-config-version.cmake`

```
spack install hsakmt-roct
```

**Comment by srekolam** [comment]
@mwkrentel yes, i have built them individually with out packages.yaml, or spack.yaml and no environment.  I did not see any failures for either 4.5.0 or 4.5.2 etc..

**Comment by mwkrentel** [comment]
Ok, here's something.  `spack debug report` says:`

```
$ spack debug report
* **Spack:** 0.11.2-16735-5bc518baf9
* **Python:** 3.6.5
* **Platform:** linux-fedora26-westmere
* **Concretizer:** clingo
```

There are two things here: (1) hsakmt-roct is somehow taking its
version from the spack version.  I've seen something similar once
before, where it mattered if you were running some command inside a
git repo or not where you ended up running git from the wrong repo.

(2) the spack version reporting is wrong.  This is not spack 0.11.2.
It's from develop today (3/10/22).

So, the real questions are: (1) why does spack get the wrong version,
and (2) why does this bleed into the hsakmt-roct version?

@becker33 @scheibelp  Any ideas?

**Comment by mwkrentel** [comment]
@srekolam  I found out what's going on, and I can show you how to reproduce the problem.

In the hsakmt-roct source tree, it uses git to identify its version in
cmake_modules/utils.cmake:

```
function ( get_version DEFAULT_VERSION_STRING )
...
    execute_process ( COMMAND git describe --tags --dirty --long
                      WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
                      OUTPUT_VARIABLE GIT_TAG_STRING
                      OUTPUT_STRIP_TRAILING_WHITESPACE
                      RESULT_VARIABLE RESULT )
```

This is dangerous because you don't know if this code is being run
from a git repo or not.

In `config.yaml`, spack uses these paths to locate the build directory:

```
build_stage:
  - $tempdir/$user/spack-stage
  - $user_cache_path/stage
# - $spack/var/spack/stage
```

If you use the first choice, outside of the spack repo, then git will
fail and the version will fall back to some default.  BUT, if you
uncomment and use the third choice, then the build directory will be
inside the spack repo, not the rocm repo.  So, git will succeed, but
it will get the spack version, not the hsakmt-roct version.

This actually bit me with hpctoolkit a couple years ago.

I suggest as a solution, you should check that the top-level source
directory after you unpack the tar file (or git checkout) must have a
`.git` directory, or else don't use git and fall back to the builtin
default.

It's an odd combination of events, but it happens to hit my use case.
It's odd, but IMO a definite bug that should be fixed both in the ROCM
repo and worked around in spack.

Ok?

----------

So, here's how to reproduce:

1. fresh git clone

2. edit etc/spack/defaults/config.yaml to put build_stage inside the
   spack repo as above

3. spack install hsakmt-roct

4. look at hsakmt-config-version.cmake for PACKAGE_VERSION

**Comment by jewelkm** [comment]
@mwkrentel Can you please confirm if you still face this issue.
I checked with ROCm 5.2.0 and we do not seeing this issue happening.

**Comment by mwkrentel** [comment]
Hmm, I just tried `spack install hip @5.2.3` and yes, the build still
fails for me in `hsa-rocr-dev` with the same error as before.

```
1 error found in build log:
     12    -- Check for working CXX compiler: /home/krentel/hsa-bug/spack-repo/lib/spack/env/gcc/g++ -
 skipped
     13    -- Detecting CXX compile features
     14    -- Detecting CXX compile features - done
     15    -- Found LibElf: /home/krentel/hsa-bug/install/linux-rhel8-broadwell/gcc-8.5.0/elfutils-0.186-bokjoeae2lywytsm
           afn3z2xpsnpi7js2/lib/libelf.so
     16    -- Performing Test ELF_GETSHDRSTRNDX
     17    -- Performing Test ELF_GETSHDRSTRNDX - Success
  >> 18    CMake Error at CMakeLists.txt:99 (find_package):
     19      Could not find a configuration file for package "hsakmt" that is compatible
     20      with requested version "1.0".
     21    
     22      The following configuration files were considered but not accepted:
     23    
     24        /home/krentel/hsa-bug/install/linux-rhel8-broadwell/gcc-8.5.0/hsakmt-roct-5.2.3-2o6gndd5vkq7lj6gy2mpdgjnii
           ons276/lib64/cmake/hsakmt/hsakmt-config.cmake, version: 0.17.1-3998-g9c1d6da111
```

Let me check the tar files, that will tell me if anything has changed or not.
